Lite Client Spec (#3796)
tendermint/tendermint: docs/spec/consensus/light-client.md

# Lite Client

A lite client is a process that connects to the Tendermint Full Node(s) and then tries to verify application data using the Merkle proofs. In this document we describe mechanisms that ensures that the Tendermint lite client
has the same level of security as Full Node processes (without being itself a Full Node).

## Lite client requirements from Tendermint and Proof of Stake modules

> Before explaining lite client mechanisms and operations we need to define some requirements expected from the Tendermint blockchain. Tendermint provides a deterministic, Byzantine fault-tolerant, source of time (called

Reviewer: So I don't think that BFT has a huge amount to with lite clients.

Weak Subjectivity is entirely in terms of clients local clock. 

It is slightly relevant in an IBC context where BFT time is the local clock because the lite client is another blockchain.

Reviewer 2: I think we need to constrain drifts between local time and BFT time to be able to come up with some guarantees. Otherwise (at least in theory) clocks can drift arbitrarily.

Reviewer 3: I thought we're going to remove BFT time, no? https://github.com/tendermint/tendermint/issues/2840

Reviewer: The plan isn't to remove BFT time but to make it proposer based so the attack is 2f+1 instead of f+1 to accelerate time.

[BFT Time](/Users/zarkomilosevic/go-workspace/src/github.com/tendermint/tendermint/docs/spec/consensus/bft-time.md)).

> BFT time is monotonically increasing and in case of at most 1/3 of voting power equivalent of faulty validators guaranteed to be close to the wall time of correct validators. For correct functioning of lite client we need a guarantee that BFT time does not drift more than some known parameter BFT_TIME_DRIFT_BOUND (that should normally be measured in hours, maybe even days) from client wall time. Note that this requirement currently only holds in case 

Reviewer 2: It is made more concrete below.

- trusted validator set hash.

Given this information, light client initialization logic will call `Validators` method on the node it is connected to 
obtain `ResultValidators` that contains validators that has committed the block h. Then we check if MerkleRoot
of the validator set is equal to the trusted validator set hash. If verification failed, initialization exits with error, otherwise it proceeds. 

Next step is determining if the block at hight h is correctly signed by the obtained validator set. This is achieved by 
calling `Commit` method for a given height that returns `SignedHeader` that is a header together with the set of 
`Precommit` messages that commit the corresponding block. Light client logic then needs to verify if the set of signatures defined the valid commit for the given block. If this verification fails, initialization exits with error, otherwise it proceeds. 

Final step (this should actually be part of the second step) of the initialization procedure is ensuring that we operate within UNBONDING_PERIOD of the trusted validator set. This can be expressed with the following formula:

`header.Time + UNBONDING_PERIOD <= Now() - BFT_TIME_DRIFT_BOUND`.

> Note that outside this time window lite client cannot trust validator set as validators could potentially unbonded its stake so security of the lite client does not hold as they are not slashable for its actions. Therefore, they can eclipse client and cheat about the system state without risk of being punished for such misbehavior. 

Reviewer 4: ```suggestion
Note that outside this time window lite client cannot trust validator set as validators could potentially have unbonded their stake so security of the lite client does not hold as they are not slashable for its actions. Therefore, they can eclipse client and cheat about the system state without risk of being punished for such misbehavior. 
```

Reviewer 4: I don't understand what eclipsing and the unbonding period have to do with one another. If the light client is eclipsed for the full duration of the unbonding period it doesn't matter whether the validators are unbonding or not, since the light client can't submit evidence anyways. Outside the unbonding period, it doesn't matter if the client is eclipsed since the state machine will reject the evidence even if it is submitted.

Reviewer 2: I assume here that being eclipsed is not permanent state of things as at any point in time you can decide to change full node you are connected to or decide to connect to new full node. In that sense there is a difference whether you are operating within unbonding period as you still have a chance of connected to correct node, detecting fork (for example) and submit evidence, compared to the case where you are outside unbonding period where there is no help if you are cheated.

Note that this formula shows a fundamental dependence of lite client security on the wall time. If UNBONDING_PERIOD

> would be defined only in terms of logical time (block heights), lite client will not have a way to know if trusted validator set is still withing its UNBONDING_PERIOD as it does not have a way of reliably determining top of the chain.

Reviewer 4: I think we should do this! (define unbonding period as a minimum of time passed *and* of blocks passed)

It's not very hard at all.

Reviewer 2: I feel the same. I think complexity is significantly added and we increase security of lite client significantly.

> Therefore, it seems that having BFT time in sync with standard notions of time (for example NTP) is necessarily for correct operations of the system. 

Reviewer 4: ```suggestion
Therefore, it seems that having BFT time in sync with standard notions of time (for example NTP) is necessary for correct operations of the system. 
```

(and see above comment)

> Lite client security depends also on the guarantee that faulty validator behavior will be punished. Therefore if a client detect faulty behavior we need to guarantee that proof of misbehavior evidence transaction will be committed within UNBONDING_PERIOD of faulty validators so it can be slashed. This can be achieved by having client considering

Reviewer 4: ```suggestion
Lite client security depends also on the guarantee that faulty validator behavior will be punished. Therefore if a client detects faulty behavior we need to guarantee that proof of misbehavior evidence transaction will be committed within UNBONDING_PERIOD of faulty validators so it can be slashed. This can be achieved by having client considering
```

  NextValSet        []Validator        // next validator set
  NextValSetHash    []byte             // hash of the next validator set
  Height            int64              // current height
  ValSetTime        int64              // time when the current and next validator set is initialized
}
```

Initialization procedure is captured by the following pseudo code:

```golang
Init(height int64, valSetHash Hash): (LiteClientState, error) {
  state = LiteClientState {}
  valSet = Validators(height).Validators
  if Hash(valSet) != valSetHash then return (nil, error)

  signedHeader = Commit(height)
  if signedHeader.Header.Time + LITE_CLIENT_CERTIFICATE_PERIOD > Now() - BFT_TIME_DRIFT_BOUND then return (nil, error)
  if votingPower(signedHeader.Commit) <= 2/3 totalVotingPower(valSet) then return (nil, error)
  
  nextValSet = Validators(height+1).Validators
  if Hash(nextValSet) != signedHeader.Header.NextValSetHash then return (nil, error)

  state.ValSet = nextValSet
  state.ValSetHash = signedHeader.Header.NextValSetHash
  state.Height = height
  state.ValSetTime = signedHeader.Header.Time
  
  return state, nil
}  
```

> To be able to validate a Merkle proof, a light client needs to validate the blockchain header that contains the root app hash.Validating a blockchain header in Tendermint consists in verifying that the header is committed (signed) by >2/3 of the voting power of the corresponding validator set. As the validator set is a dynamic set (it is changing), one of the core functionality of the lite client is updating the current validator set, that is then used to verify the

Reviewer 4: ```suggestion
To be able to validate a Merkle proof, a light client needs to validate the blockchain header that contains the root app hash. Validating a blockchain header in Tendermint consists in verifying that the header is committed (signed) by >2/3 of the voting power of the corresponding validator set. As the validator set is a dynamic set (it is changing), one of the core functionalities of the lite client is updating the current validator set, which is then used to verify the
```

> blockchain header, and further the corresponding Merkle proofs.

Reviewer 4: ```suggestion
blockchain header, and subsequently the corresponding Merkle proofs.
```

  if state.ValSetTime + LITE_CLIENT_CERTIFICATE_PERIOD > Now() - BFT_TIME_DRIFT_BOUND then return error

  for i = state.Height+1; i < height; i++ {
    signedHeader = Commit(i)
    if votingPower(signedHeader.Commit) <= 2/3 totalVotingPower(state.ValSet) then return error
    if signedHeader.Header.ValidatorsHash != state.ValSetHash then return error

    nextValSetHash = signedHeader.Header.NextValidatorsHash

    nextValSet = Validators(i+1).Validators
    if Hash(nextValSet) != nextValSetHash then return (0, error)

    // at this point we can install new valset
    state.ValSet = nextValSet
    state.ValSetHash = nextValSetHash
    state.Height = i
    state.ValSetTime = signedHeader.Header.Time
  }  
}

Note that in the logic above we assume that the lite client will always go upward with respect to header verifications,
i.e., that it will always be used to verify more recent headers. Going backward is problematic as having trust in 
some validator set at height h does not give us trust in validator set before height h. Therefore, if lite client query
is about smaller height, trusted validator set for smaller heights is needed. 

> In case a call to the FullNode or subsequent checks fail, a light client need to implement some recovery strategy, for example connecting to other FullNode.

Reviewer 4: The light client should always try to use a randomized load balancing strategy, considering the possibility of malevolent eclipse attacks (or just innocuous but inconvenient stale full node data).
